Regression in org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor

PR #976 creates this exception
 ```
10:09:08 ERROR: Error during SonarQube Scanner execution
10:09:08 org.sonar.squidbridge.api.AnalysisException: Unable to analyze file: T:\xxxx\Src\xxxx.h
10:09:08         at org.sonar.squidbridge.AstScanner.scanFiles(AstScanner.java:131)
10:09:08         at org.sonar.plugins.cxx.squid.CxxSquidSensor.execute(CxxSquidSensor.java:128)
10:09:08         at org.sonar.batch.sensor.SensorWrapper.analyse(SensorWrapper.java:57)
10:09:08         at org.sonar.batch.phases.SensorsExecutor.executeSensor(SensorsExecutor.java:58)
10:09:08         at org.sonar.batch.phases.SensorsExecutor.execute(SensorsExecutor.java:50)
10:09:08         at org.sonar.batch.phases.AbstractPhaseExecutor.execute(AbstractPhaseExecutor.java:83)
10:09:08         at org.sonar.batch.scan.ModuleScanContainer.doAfterStart(ModuleScanContainer.java:192)
10:09:08         at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
10:09:08         at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
10:09:08         at org.sonar.batch.scan.ProjectScanContainer.scan(ProjectScanContainer.java:241)
10:09:08         at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:236)
10:09:08         at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:234)
10:09:08         at org.sonar.batch.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:226)
10:09:08         at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
10:09:08         at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
10:09:08         at org.sonar.batch.task.ScanTask.execute(ScanTask.java:47)
10:09:08         at org.sonar.batch.task.TaskContainer.doAfterStart(TaskContainer.java:86)
10:09:08         at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:142)
10:09:08         at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:127)
10:09:08         at org.sonar.batch.bootstrap.GlobalContainer.executeTask(GlobalContainer.java:106)
10:09:08         at org.sonar.batch.bootstrapper.Batch.executeTask(Batch.java:119)
10:09:08         at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:62)
10:09:08         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
10:09:08         at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
10:09:08         at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
10:09:08         at java.lang.reflect.Method.invoke(Unknown Source)
10:09:08         at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
10:09:08         at com.sun.proxy.$Proxy0.execute(Unknown Source)
10:09:08         at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:240)
10:09:08         at org.sonarsource.scanner.api.EmbeddedScanner.runAnalysis(EmbeddedScanner.java:151)
10:09:08         at org.sonarsource.scanner.cli.Main.runAnalysis(Main.java:110)
10:09:08         at org.sonarsource.scanner.cli.Main.execute(Main.java:72)
10:09:08         at org.sonarsource.scanner.cli.Main.main(Main.java:60)
10:09:08 Caused by: java.lang.ClassCastException: com.sonar.sslr.api.GenericTokenType cannot be cast to org.sonar.cxx.api.CxxKeyword
10:09:08         at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.isTypedef(AbstractCxxPublicApiVisitor.java:244)
10:09:08         at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitDeclaratorList(AbstractCxxPublicApiVisitor.java:208)
10:09:08         at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitNode(AbstractCxxPublicApiVisitor.java:161)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.visitNode(AstWalker.java:114)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:85)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.visitChildren(AstWalker.java:99)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.visit(AstWalker.java:87)
10:09:08         at com.sonar.sslr.impl.ast.AstWalker.walkAndVisit(AstWalker.java:69)
10:09:08         at org.sonar.squidbridge.AstScanner.scanFiles(AstScanner.java:110)
10:09:08         ... 32 more
```

Hi @Bertk,

It's a cast error in `isTypedef`, I will fix it. Would be interesting to know the related source code. Must be something with `typedef ...`.
```
Caused by: java.lang.ClassCastException: com.sonar.sslr.api.GenericTokenType cannot be cast to org.sonar.cxx.api.CxxKeyword
10:09:08         at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.isTypedef(AbstractCxxPublicApiVisitor.java:244)
10:09:08         at org.sonar.cxx.visitors.AbstractCxxPublicApiVisitor.visitDeclaratorList(AbstractCxxPublicApiVisitor.java:208)
```
Regards,

--- a/cxx-squid/src/main/java/org/sonar/cxx/visitors/AbstractCxxPublicApiVisitor.java
+++ b/cxx-squid/src/main/java/org/sonar/cxx/visitors/AbstractCxxPublicApiVisitor.java
@@ -34,6 +34,7 @@ import com.sonar.sslr.api.AstNode;
 import com.sonar.sslr.api.GenericTokenType;
 import com.sonar.sslr.api.Grammar;
 import com.sonar.sslr.api.Token;
+import com.sonar.sslr.api.TokenType;
 import com.sonar.sslr.api.Trivia;
 import com.sonar.sslr.impl.ast.AstXmlPrinter;
 
@@ -241,12 +242,12 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
       if (firstDeclSpecifier != null && firstDeclSpecifier.getToken().getType() == CxxKeyword.TYPEDEF) {
         AstNode classSpefifier = firstDeclSpecifier.getNextSibling();
         if (classSpefifier != null) {
-          switch ((CxxKeyword) classSpefifier.getToken().getType()) {
-            case STRUCT:
-            case CLASS:
-            case UNION:
-            case ENUM:
-              return true;
+          TokenType type = classSpefifier.getToken().getType();
+          if (type == CxxKeyword.STRUCT
+            || type == CxxKeyword.CLASS
+            || type == CxxKeyword.UNION
+            || type == CxxKeyword.ENUM) {
+            return true;
           }
         }
       }
@@ -774,19 +775,19 @@ public abstract class AbstractCxxPublicApiVisitor<GRAMMAR extends Grammar>
             CxxKeyword.ENUM, CxxKeyword.UNION);
 
         if (enclosingSpecifierNode != null) {
-          switch ((CxxKeyword) enclosingSpecifierNode.getToken()
-            .getType()) {
-            case STRUCT:
-            case UNION:
-              // struct and union members have public access, thus access level
-              // is the access level of the enclosing classSpecifier
-              return isPublicApiMember(classSpecifier);
-            case CLASS:
-              // default access in classes is private
-              return false;
-            default:
-              LOG.error("isPublicApiMember unhandled case: {} at {}", enclosingSpecifierNode.getType(), enclosingSpecifierNode.getTokenLine());
-              return false;
+          TokenType type = enclosingSpecifierNode.getToken().getType();
+          if (type == CxxKeyword.STRUCT || type == CxxKeyword.UNION) {
+            // struct and union members have public access, thus access level
+            // is the access level of the enclosing classSpecifier
+            return isPublicApiMember(classSpecifier);
+
+          } else if (type == CxxKeyword.CLASS) {
+            // default access in classes is private
+            return false;
+
+          } else {
+            LOG.error("isPublicApiMember unhandled case: {} at {}", enclosingSpecifierNode.getType(), enclosingSpecifierNode.getTokenLine());
+            return false;
           }
         } else {
           LOG.error("isPublicApiMember: failed to get enclosing classSpecifier for node at {}", node.getTokenLine());

